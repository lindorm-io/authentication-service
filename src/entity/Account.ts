import { AccountEvent } from "../enum";
import { EntityBase, EntityCreationError, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { Permission } from "@lindorm-io/jwt";

export interface IAccountOTP {
  signature: string;
  uri: string;
}

export interface IAccountPassword {
  signature: string;
  updated: Date;
}

export interface IAccount extends IEntity {
  email: string;
  identityLinked: boolean;
  otp: IAccountOTP;
  password: IAccountPassword;
  permission: string;
}

export interface IAccountOptions extends IEntityBaseOptions {
  email: string;
  identityLinked?: boolean;
  otp?: IAccountOTP;
  password?: IAccountPassword;
  permission?: string;
}

export class Account extends EntityBase implements IAccount {
  private _email: string;
  private _identityLinked: boolean;
  private _otp: IAccountOTP;
  private _password: IAccountPassword;
  private _permission: string;

  constructor(options: IAccountOptions) {
    super(options);
    this._email = options.email.toLowerCase();
    this._identityLinked = options.identityLinked || false;
    this._otp = {
      signature: options.otp?.signature || null,
      uri: options.otp?.uri || null,
    };
    this._password = {
      signature: options.password?.signature || null,
      updated: options.password?.updated || null,
    };
    this._permission = options.permission || Permission.USER;
  }

  public get email(): string {
    return this._email;
  }
  public set email(email: string) {
    this._email = email.toLowerCase();
    this.addEvent(AccountEvent.EMAIL_CHANGED, { email: this._email });
  }

  public get identityLinked(): boolean {
    return this._identityLinked;
  }
  public set identityLinked(identityLinked: boolean) {
    this._identityLinked = identityLinked;
    this.addEvent(AccountEvent.EMAIL_CHANGED, { identityLinked: this._identityLinked });
  }

  public get otp(): IAccountOTP {
    return this._otp;
  }
  public set otp(otp: IAccountOTP) {
    this._otp = otp;
    this.addEvent(AccountEvent.OTP_CHANGED, { otp: this._otp });
  }

  public get password(): IAccountPassword {
    return this._password;
  }
  public set password(password: IAccountPassword) {
    this._password = password;
    this.addEvent(AccountEvent.PASSWORD_CHANGED, { password: this._password });
  }

  public get permission(): string {
    return this._permission;
  }
  public set permission(permission: string) {
    this._permission = permission;
    this.addEvent(AccountEvent.PERMISSION_CHANGED, { permission: this._permission });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== AccountEvent.CREATED) continue;
      throw new EntityCreationError("Account");
    }
    this.addEvent(AccountEvent.CREATED, {
      email: this._email,
      permission: this._permission,
      password: this._password,
      otp: this._otp,
      created: this._created,
      updated: this._updated,
    });
  }
}
