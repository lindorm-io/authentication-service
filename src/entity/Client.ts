import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { ClientEvent } from "../enum";

export interface IClient extends IEntity {
  approved: boolean;
  description: string;
  emailAuthorizationUri: string;
  name: string;
  secret: string;
}

export interface IClientOptions extends IEntityBaseOptions {
  approved?: boolean;
  description?: string;
  emailAuthorizationUri?: string;
  name?: string;
  secret?: string;
}

export class Client extends EntityBase implements IClient {
  private _approved: boolean;
  private _description: string;
  private _emailAuthorizationUri: string;
  private _name: string;
  private _secret: string;

  constructor(options?: IClientOptions) {
    super(options);
    this._approved = options?.approved || false;
    this._description = options?.description || null;
    this._emailAuthorizationUri = options?.emailAuthorizationUri || null;
    this._name = options?.name || null;
    this._secret = options?.secret || null;
  }

  public get approved(): boolean {
    return this._approved;
  }
  public set approved(approved: boolean) {
    this._approved = approved;
    this.addEvent(ClientEvent.APPROVED_CHANGED, { approved: this._approved });
  }

  public get emailAuthorizationUri(): string {
    return this._emailAuthorizationUri;
  }
  public set emailAuthorizationUri(uri: string) {
    this._emailAuthorizationUri = uri;
    this.addEvent(ClientEvent.EMAIL_AUTHORIZATION_URI_CHANGED, { uri: this._emailAuthorizationUri });
  }

  public get description(): string {
    return this._description;
  }
  public set description(description: string) {
    this._description = description;
    this.addEvent(ClientEvent.DESCRIPTION_CHANGED, { description: this._description });
  }

  public get name(): string {
    return this._name;
  }
  public set name(name: string) {
    this._name = name;
    this.addEvent(ClientEvent.NAME_CHANGED, { name: this._name });
  }

  public get secret(): string {
    return this._secret;
  }
  public set secret(secret: string) {
    this._secret = secret;
    this.addEvent(ClientEvent.SECRET_CHANGED, { secret: this._secret });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== ClientEvent.CREATED) continue;
      throw new Error("Client has already been created");
    }
    this.addEvent(ClientEvent.CREATED, {
      name: this._name,
      emailAuthorizationUri: this._emailAuthorizationUri,
      approved: this._approved,
      description: this._description,
      secret: this._secret,
      created: this._created,
      updated: this._updated,
    });
  }
}
