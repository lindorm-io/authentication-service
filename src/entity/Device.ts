import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { DeviceEvent } from "../enum";

export interface IDevicePIN {
  signature: string;
  updated: Date;
}

export interface IDevice extends IEntity {
  accountId: string;
  name: string;
  pin: IDevicePIN;
  publicKey: string;
  secret: string;
}

export interface IDeviceOptions extends IEntityBaseOptions {
  accountId: string;
  name?: string;
  pin?: IDevicePIN;
  publicKey: string;
  secret?: string;
}

export class Device extends EntityBase implements IDevice {
  private _accountId: string;
  private _name: string;
  private _pin: IDevicePIN;
  private _publicKey: string;
  private _secret: string;

  constructor(options: IDeviceOptions) {
    super(options);
    this._accountId = options.accountId;
    this._name = options.name || null;
    this._pin = {
      signature: options.pin?.signature || null,
      updated: options.pin?.updated || null,
    };
    this._publicKey = options.publicKey;
    this._secret = options.secret || null;
  }

  public get accountId(): string {
    return this._accountId;
  }

  public get name(): string {
    return this._name;
  }
  public set name(name: string) {
    this._name = name;
    this.addEvent(DeviceEvent.NAME_CHANGED, { name: this._name });
  }

  public get pin(): IDevicePIN {
    return this._pin;
  }
  public set pin(pin: IDevicePIN) {
    this._pin = pin;
    this.addEvent(DeviceEvent.PIN_CHANGED, { pin: this._pin });
  }

  public get publicKey(): string {
    return this._publicKey;
  }

  public get secret(): string {
    return this._secret;
  }
  public set secret(secret: string) {
    this._secret = secret;
    this.addEvent(DeviceEvent.SECRET_CHANGED, { secret: this._secret });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== DeviceEvent.CREATED) continue;
      throw new Error("Device has already been created");
    }
    this.addEvent(DeviceEvent.CREATED, {
      accountId: this._accountId,
      name: this._name,
      pin: this._pin,
      publicKey: this._publicKey,
      secret: this._secret,
      created: this._created,
      updated: this._updated,
    });
  }
}
