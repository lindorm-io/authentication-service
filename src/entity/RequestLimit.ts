import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { GrantType } from "../enum";

export interface IRequestLimit extends IEntity {
  backOffUntil: Date;
  failedTries: number;
  grantType: GrantType;
  subject: string;
}

export interface IRequestLimitOptions extends IEntityBaseOptions {
  backOffUntil?: Date;
  failedTries?: number;
  grantType: GrantType;
  subject: string;
}

export class RequestLimit extends EntityBase implements IRequestLimit {
  private _backOffUntil: Date;
  private _failedTries: number;
  private _grantType: GrantType;
  private _subject: string;

  constructor(options: IRequestLimitOptions) {
    super(options);
    this._backOffUntil = options.backOffUntil || null;
    this._failedTries = options.failedTries || 1;
    this._grantType = options.grantType;
    this._subject = options.subject;
  }

  public get backOffUntil(): Date {
    return this._backOffUntil;
  }
  public set backOffUntil(backOffUntil: Date) {
    this._backOffUntil = backOffUntil;
    this._updated = new Date();
  }

  public get failedTries(): number {
    return this._failedTries;
  }
  public set failedTries(failedTries: number) {
    this._failedTries = failedTries;
    this._updated = new Date();
  }

  public get grantType(): GrantType {
    return this._grantType;
  }

  public get subject(): string {
    return this._subject;
  }

  public create(): void {
    /* intentionally left empty */
  }
}
