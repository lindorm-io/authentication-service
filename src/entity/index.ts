export * from "./Account";
export * from "./Client";
export * from "./Device";
export * from "./Session";
