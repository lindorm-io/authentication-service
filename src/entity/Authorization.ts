import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/entity";
import { GrantType } from "../enum";
import { Scope } from "@lindorm-io/jwt";

export interface IAuthorization extends IEntity {
  challengeId: string;
  clientId: string;
  codeChallenge: string;
  codeMethod: string;
  deviceId: string;
  email: string;
  expires: Date;
  grantType: GrantType;
  otpCode: string;
  redirectUri: string;
  responseType: string;
  scope: Array<Scope>;
}

export interface IAuthorizationOptions extends IEntityBaseOptions {
  challengeId?: string;
  clientId: string;
  codeChallenge: string;
  codeMethod: string;
  deviceId?: string;
  email: string;
  expires: Date;
  grantType: GrantType;
  otpCode?: string;
  redirectUri: string;
  responseType: string;
  scope: Array<Scope>;
}

export class Authorization extends EntityBase implements IAuthorization {
  readonly challengeId: string;
  readonly clientId: string;
  readonly codeChallenge: string;
  readonly codeMethod: string;
  readonly deviceId: string;
  readonly email: string;
  readonly expires: Date;
  readonly grantType: GrantType;
  readonly otpCode: string;
  readonly redirectUri: string;
  readonly responseType: string;
  readonly scope: Array<Scope>;

  constructor(options: IAuthorizationOptions) {
    super(options);

    this.challengeId = options.challengeId || null;
    this.clientId = options.clientId;
    this.codeChallenge = options.codeChallenge;
    this.codeMethod = options.codeMethod;
    this.deviceId = options.deviceId || null;
    this.email = options.email;
    this.expires = options.expires;
    this.grantType = options.grantType;
    this.otpCode = options.otpCode || null;
    this.redirectUri = options.redirectUri;
    this.responseType = options.responseType;
    this.scope = options.scope;
  }

  public create(): void {
    /* intentionally left empty */
  }
}
