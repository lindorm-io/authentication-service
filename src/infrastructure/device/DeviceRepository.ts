import { Device, IDevice } from "../../entity";
import { IRepository, IRepositoryOptions, RepositoryBase } from "@lindorm-io/mongo";
import { MongoCollection } from "../../enum";
import { indices } from "./indices";
import { schema } from "./schema";

export interface IDeviceFilter {
  id?: string;
  accountId?: string;
  name?: string;
}

export interface IDeviceRepository extends IRepository<Device> {
  create(entity: Device): Promise<Device>;
  update(entity: Device): Promise<Device>;
  find(filter: IDeviceFilter): Promise<Device>;
  findMany(filter: IDeviceFilter): Promise<Array<Device>>;
  findOrCreate(filter: IDeviceFilter): Promise<Device>;
  remove(entity: Device): Promise<void>;
}

export class DeviceRepository extends RepositoryBase<Device> implements IDeviceRepository {
  constructor(options: IRepositoryOptions) {
    super({
      collectionName: MongoCollection.DEVICE,
      db: options.db,
      logger: options.logger,
      indices,
      schema,
    });
  }

  protected createEntity(data: IDevice): Device {
    return new Device(data);
  }

  protected getEntityJSON(entity: Device): IDevice {
    return {
      id: entity.id,
      version: entity.version,
      created: entity.created,
      updated: entity.updated,
      events: entity.events,

      accountId: entity.accountId,
      name: entity.name,
      pin: entity.pin,
      publicKey: entity.publicKey,
      secret: entity.secret,
    };
  }

  public async create(entity: Device): Promise<Device> {
    return super.create(entity);
  }

  public async update(entity: Device): Promise<Device> {
    return super.update(entity);
  }

  public async find(filter: IDeviceFilter): Promise<Device> {
    return super.find(filter);
  }

  public async findMany(filter: IDeviceFilter): Promise<Array<Device>> {
    return super.findMany(filter);
  }

  public async remove(entity: Device): Promise<void> {
    await super.remove(entity);
  }
}
