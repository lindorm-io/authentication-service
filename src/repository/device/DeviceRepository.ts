import { Device, IDevice } from "../../entity";
import { IRepository, IRepositoryOptions, RepositoryBase } from "@lindorm-io/mongo";
import { MongoCollection } from "../../enum";
import { indices } from "./DeviceRepository.indices";
import { schema } from "./DeviceRepository.schema";

export interface IDeviceFilter {
  id?: string;
  accountId?: string;
  name?: string;
}

export interface IDeviceRepository extends IRepository<Device> {
  create(device: Device): Promise<Device>;
  update(device: Device): Promise<Device>;
  find(filter: IDeviceFilter): Promise<Device>;
  findMany(filter: IDeviceFilter): Promise<Array<Device>>;
  remove(device: Device): Promise<void>;
}

export class DeviceRepository extends RepositoryBase<Device> implements IDeviceRepository {
  constructor(options: IRepositoryOptions) {
    super({
      collectionName: MongoCollection.DEVICE,
      db: options.db,
      logger: options.logger,
      indices,
      schema,
    });
  }

  protected createEntity(data: IDevice): Device {
    return new Device(data);
  }

  protected getEntityJSON(device: Device): IDevice {
    return {
      id: device.id,
      version: device.version,
      created: device.created,
      updated: device.updated,
      events: device.events,

      accountId: device.accountId,
      name: device.name,
      pin: device.pin,
      publicKey: device.publicKey,
      secret: device.secret,
    };
  }

  public async create(device: Device): Promise<Device> {
    return super.create(device);
  }

  public async update(device: Device): Promise<Device> {
    return super.update(device);
  }

  public async find(filter: IDeviceFilter): Promise<Device> {
    return super.find(filter);
  }

  public async findMany(filter: IDeviceFilter): Promise<Array<Device>> {
    return super.findMany(filter);
  }

  public async remove(device: Device): Promise<void> {
    await super.remove(device);
  }
}
