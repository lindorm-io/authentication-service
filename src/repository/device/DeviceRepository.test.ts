import MockDate from "mockdate";
import { Device } from "../../entity";
import { DeviceRepository } from "./DeviceRepository";
import { MOCK_LOGGER } from "../../test/mocks/logger";
import { MOCK_UUID } from "../../test/mocks/data";
import { MongoInMemoryConnection, RepositoryEntityNotFoundError } from "@lindorm-io/mongo";
import { baseHash } from "@lindorm-io/core";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("DeviceRepository", () => {
  let repository: DeviceRepository;
  let device: Device;

  beforeEach(() => {
    const mongo = new MongoInMemoryConnection({
      host: "host",
      name: "name",
      password: "password",
      port: 999,
      user: "user",
    });
    repository = new DeviceRepository({
      db: mongo.db(),
      logger: MOCK_LOGGER,
    });
    device = new Device({
      id: MOCK_UUID,

      accountId: MOCK_UUID,
      name: "name",
      pin: {
        signature: baseHash("signature"),
        updated: new Date(),
      },
      publicKey: "publicKey",
      secret: baseHash("secret"),
    });
  });

  test("should create", async () => {
    await expect(repository.create(device)).resolves.toMatchSnapshot();
  });

  test("should update", async () => {
    await repository.create(device);

    await expect(repository.update(device)).resolves.toMatchSnapshot();
  });

  test("should find", async () => {
    await repository.create(device);

    await expect(repository.find({ name: "name" })).resolves.toMatchSnapshot();
  });

  test("should find many", async () => {
    await repository.create(device);
    await repository.create(
      new Device({
        accountId: MOCK_UUID,
        name: "other",
        publicKey: "other-publicKey",
      }),
    );

    await expect(repository.findMany({ accountId: MOCK_UUID })).resolves.toMatchSnapshot();
  });

  test("should remove", async () => {
    await repository.create(device);

    await expect(repository.remove(device)).resolves.toBe(undefined);
    await expect(repository.find({ id: device.id })).rejects.toStrictEqual(expect.any(RepositoryEntityNotFoundError));
  });
});
