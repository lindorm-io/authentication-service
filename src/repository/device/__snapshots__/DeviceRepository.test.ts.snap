// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeviceRepository should create 1`] = `
Device {
  "_accountId": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [],
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_name": "name",
  "_pin": Object {
    "signature": "c2lnbmF0dXJl",
    "updated": 2020-01-01T07:00:00.000Z,
  },
  "_publicKey": "publicKey",
  "_secret": "c2VjcmV0",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`DeviceRepository should find 1`] = `
Device {
  "_accountId": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [],
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_name": "name",
  "_pin": Object {
    "signature": "c2lnbmF0dXJl",
    "updated": 2020-01-01T07:00:00.000Z,
  },
  "_publicKey": "publicKey",
  "_secret": "c2VjcmV0",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
}
`;

exports[`DeviceRepository should find many 1`] = `
Array [
  Device {
    "_accountId": "be3a62d1-24a0-401c-96dd-3aff95356811",
    "_created": 2020-01-01T07:00:00.000Z,
    "_events": Array [],
    "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
    "_name": "name",
    "_pin": Object {
      "signature": "c2lnbmF0dXJl",
      "updated": 2020-01-01T07:00:00.000Z,
    },
    "_publicKey": "publicKey",
    "_secret": "c2VjcmV0",
    "_updated": 2020-01-01T07:00:00.000Z,
    "_version": 0,
  },
  Device {
    "_accountId": "be3a62d1-24a0-401c-96dd-3aff95356811",
    "_created": 2020-01-01T07:00:00.000Z,
    "_events": Array [],
    "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
    "_name": "other",
    "_pin": Object {
      "signature": null,
      "updated": null,
    },
    "_publicKey": "other-publicKey",
    "_secret": null,
    "_updated": 2020-01-01T07:00:00.000Z,
    "_version": 0,
  },
]
`;

exports[`DeviceRepository should update 1`] = `
Device {
  "_accountId": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [],
  "_id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "_name": "name",
  "_pin": Object {
    "signature": "c2lnbmF0dXJl",
    "updated": 2020-01-01T07:00:00.000Z,
  },
  "_publicKey": "publicKey",
  "_secret": "c2VjcmV0",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 1,
}
`;
